llama.cpp

```diff
@@ -8004,7 +8004,13 @@ void llama_sample_top_k(struct llama_context * ctx, llama_token_data_array * can
         if (k == (int) candidates->size) {
```

Comment:
You probably want to change this to `k >= (int) (3*candidates->size()/4)` or similar. At least on my computer, just `std::sort()` is faster than `std::nth_element() + std::partial_sort()` for more than 24,000 tokens or so when `candidates->size` is 32,000.

Interestingly enough, on my computer (Ubuntu 22.04, GCC 11.4.0, Ryzen-5975WX), `nth_element + partial_sort` is faster than just `partial_sort` above about 1500 tokens, but below the two are identical. But than again, the time I measure for sorting in a stand alone benchmark is ~2X shorter than the sampling times I see reported in the comments, even though the Ryzen-5975WX doesn't have the best single-thread performance.

Reply:
I think the lower border is CPU and possibly platform dependent.
For me partial_sort is significantly faster than nth-sort() for low top-k (like 1-500), the advantage shrinks and they balance at around 3000, then it turns.
Looks like at the high end sort() again wins out over both of them.

I'll add your suggestion to the sort() branch, I'm also testing a min_p change

Reply:
> You probably want to change this to `k >= (int) (3*candidates->size()/4)` or similar. At least on my computer, just `std::sort()` is faster than `std::nth_element() + std::partial_sort()` for more than 24,000 tokens or so when `candidates->size` is 32,000.
> 
> Interestingly enough, on my computer (Ubuntu 22.04, GCC 11.4.0, Ryzen-5975WX), `nth_element + partial_sort` is faster than just `partial_sort` above about 1500 tokens, but below the two are identical. But than again, the time I measure for sorting in a stand alone benchmark is ~2X shorter than the sampling times I see reported in the comments, even though the Ryzen-5975WX doesn't have the best single-thread performance.

Now that I think about it, I'm not sure if a dynamic limit like that is the best choice.
It's probably better to have an absolute number in there.
For Falcon or chinese fine tunes the vocabulary can be huge, so that 3/4 size switch comes very late.